runtime/doc/lua.txt

```diff
@@ -4429,4 +4429,216 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
         (`string[]`)
 
 
+==============================================================================
+Lua module: vim.net                                                  *vim.net*
+
+*HeaderTable*
+
+    Fields: ~
+      • {_storage}  (`table<string, string[]>`)
+      • {set}       (`fun(self: HeaderTableself: HeaderTablevalue: string[]|stringkey: string)`)
+                    Set value of header.
+      • {append}    (`fun(self: HeaderTableself: HeaderTablekey: string)`)
+                    Append value to header.
+      • {get}       (`fun(self: HeaderTableself: HeaderTablekey: string): string?`)
+                    Get header value. For headers like cookies, use
+                    |HeaderTable:get_all|.
+      • {get_all}   (`fun(self: HeaderTableself: HeaderTablekey: string): string[]?`)
+                    Get all header values.
+      • {has}       (`fun(self: HeaderTableself: HeaderTablekey: string): boolean`)
+
+*vim.net.createCurlArgs.Opts*
+
+    Fields: ~
+      • {method}             (`vim.net.httpMethod?`)
+      • {redirect}           (`"follow"|"error"?`)
+      • {data}               (`string|table<string, any>?`)
+      • {headers}            (`HeaderTable|table<string, string|string[]>?`)
+      • {download_location}  (`string?`)
+      • {upload_file}        (`string?`)
+      • {user}               (`string?`)
+
+
+HeaderTable:append({key})                               *HeaderTable:append()*
+    Append value to header.
+
+    Parameters: ~
+      • {key}  (`string`) Non case-sensitive header name.
+
+HeaderTable:get({key})                                     *HeaderTable:get()*
+    Get header value. For headers like cookies, use |HeaderTable:get_all|.
+
+    Parameters: ~
+      • {key}  (`string`) Non case-sensitive header name.
+
+    Return: ~
+        (`string?`)
+
+HeaderTable:get_all({key})                             *HeaderTable:get_all()*
+    Get all header values.
+
+    Parameters: ~
+      • {key}  (`string`) Non case-sensitive header name.
+
+    Return: ~
+        (`string[]?`)
+
+HeaderTable:has({key})                                     *HeaderTable:has()*
+
+    Parameters: ~
+      • {key}  (`string`) Non case-sensitive header name.
+
+    Return: ~
+        (`boolean`) has true if the HeaderTable contains key.
+
+HeaderTable:set({value}, {key})                            *HeaderTable:set()*
+    Set value of header.
+
+    Parameters: ~
+      • {value}  (`string[]|string`) Header value.
+      • {key}    (`string`) Non case-sensitive header name.
+
+vim.net.download({url}, {path}, {opts})                   *vim.net.download()*
+    Asynchronously download a file. To read the response metadata, such as
+    headers and body, use |vim.net.fetch()|.
+
+    See man://curl for supported protocols. Not all protocols are fully
+    tested.
+
+    Please carefully note the option differences with |vim.net.fetch()|,
```

Comment:
what are the differences? at a glance doesn't look like they need to be separate functions.

as clason also mentioned, we could start with a minimal version of `vim.net.fetch` which has the interface (signature) that we want, but only supports basic "download" behavior.

Reply:
Fetch is _much_ more general, allowing arbitrary requests (including _writing_, needed for `:e`). Download is a simplified interface for just... downloading a file (but not enough to implement `:e` over network). But note that you still need to expose some curl options since network conditions are very heterogeneous.

So both are needed (if both of these different features are wanted). I agree the name `fetch` is bad, though, and should be changed to make that clearer.

Reply:
> Fetch is _much_ more general,

why can't that be an overload? I still don't get why we need two different functions